clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -1011,8 +996,13 @@ private Record nextFetchedRecord() {
                         drain();
                         return null;
                     }
-                    currentBatch = batches.next();
-                    maybeEnsureValid(currentBatch);
+
+                    if (hasExceptionInLastFetch) {
+                        currentBatch.ensureValid();
+                    } else {
+                        currentBatch = batches.next();
+                        maybeEnsureValid(currentBatch);
```

Comment:
Should we set `lastRecord` to null here? Otherwise it seems like we might misinterpret which record had the error.